akka-http-tests/src/test/scala/akka/http/scaladsl/common/SSLContextUtilsSpec.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.common
+
+import akka.http.scaladsl.ClientTransport
+import akka.http.scaladsl.ConnectionContext
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.HttpsConnectionContext
+import akka.http.scaladsl.model.HttpRequest
+import akka.http.scaladsl.model.StatusCodes
+import akka.http.scaladsl.settings.ConnectionPoolSettings
+import akka.testkit.AkkaSpec
+import akka.testkit.TestProbe
+import com.typesafe.config.ConfigFactory
+import org.scalatest.matchers.should.Matchers
+
+import java.io.InputStream
+import java.net.InetSocketAddress
+import java.nio.file.Path
+import java.security.KeyStore
+import java.security.SecureRandom
+import javax.net.ssl.SSLContext
+import javax.net.ssl.TrustManagerFactory
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+class SSLContextUtilsSpec extends AkkaSpec with Matchers {
+  "The SSLContextUtils" should {
+    "conveniently load pem files" in {
+      SSLContextUtils.constructSSLContext(ConfigFactory.parseString(
+        """
+           my-server {
+             certificate = "akka-http-tests/src/test/resources/certs/example.com.crt"
+             private-key = "akka-http-tests/src/test/resources/certs/example.com.key"
+             ca-certificates = ["akka-http-tests/src/test/resources/certs/exampleca.crt"]
+           }
+          """).getConfig("my-server"))
+
+      SSLContextUtils.constructSSLContext(
+        Path.of("akka-http-tests/src/test/resources/certs/example.com.crt"),
+        Path.of("akka-http-tests/src/test/resources/certs/example.com.key"),
+        Seq(Path.of("akka-http-tests/src/test/resources/certs/exampleca.crt"))
+      )
+    }
+
+    "create a refreshing context provider" in {
+      val probe = TestProbe()
+      val provider = SSLContextUtils.refreshingSSLEngineProvider(10.millis) { () =>
+        probe.ref ! "Constructed"
+        SSLContextUtils.constructSSLContext(
+          Path.of("akka-http-tests/src/test/resources/certs/example.com.crt"),
+          Path.of("akka-http-tests/src/test/resources/certs/example.com.key"),
+          Seq(Path.of("akka-http-tests/src/test/resources/certs/exampleca.crt"))
+        )
+      }
+      val context1 = provider()
+      probe.expectMsg("Constructed")
+      Thread.sleep(20)
```

Comment:
There is no active scheduling kick in, but a check in each call if that much time elapsed since the cached context was provided, so this should be completely deterministic.

Reply:
In actual use of the provider, each of those calls triggered by a new incoming TCP connection, but here we just call it directly.